qir/qat/ValidationPass/ValidationPass.cpp

```diff
@@ -12,93 +12,280 @@ namespace microsoft
-                            {
-                                ++internal_calls_[name];
-                            }
-                            else
-                            {
-                                internal_calls_[name] = 1;
-                            }
-                        }
-                    }
+            if (f->isDeclaration())
+            {
+                if (external_calls_.find(name) != external_calls_.end())
+                {
+                    ++external_calls_[name];
+                }
+                else
+                {
+                    external_calls_[name] = 1;
                 }
+
+                external_call_location_[name].push_back(current_location_);
+            }
+            else
+            {
+                if (internal_calls_.find(name) != internal_calls_.end())
+                {
+                    ++internal_calls_[name];
+                }
+                else
+                {
+                    internal_calls_[name] = 1;
+                }
+
+                internal_call_location_[name].push_back(current_location_);
             }
         }
+    }
 
-        bool raise_exception = false;
+    void ValidationPass::pointerChecks(Instruction& instr)
+    {
+        for (auto& op : instr.operands())
+        {
+            // Skipping function pointers
+            auto function_pointer = llvm::dyn_cast<llvm::Function>(op);
+            if (function_pointer)
+            {
+                continue;
+            }
+
+            // Skipping non-pointer types
+            auto pointer_type = llvm::dyn_cast<llvm::PointerType>(op->getType());
+            if (!pointer_type)
+            {
+                continue;
+            }
+
+            uint64_t    n            = 1;
+            llvm::Type* element_type = pointer_type->getElementType();
+            while (pointer_type = llvm::dyn_cast<llvm::PointerType>(element_type))
+            {
+                element_type = pointer_type->getElementType();
+                ++n;
+            }
+
+            String name{};
+
+            if (element_type->isStructTy())
+            {
+                name = static_cast<String>(element_type->getStructName());
+            }
+            else
+            {
+                llvm::raw_string_ostream rso(name);
+                element_type->print(rso);
+            }
+
+            // Adding indirection to the name
+            while (n != 0)
+            {
+                name += "*";
+                --n;
+            }
+
+            auto it = pointers_.find(name);
+            if (it == pointers_.end())
+            {
+                pointers_[name] = 1;
+            }
+            else
+            {
+                ++it->second;
+            }
+
+            pointer_location_[name].push_back(current_location_);
+        }
+    }
+
+    bool ValidationPass::enforceOpcodeRequirements()
```

Comment:
I tend to prefer names for functions that return boolean to have a verb that matches the boolean intent, so that it matches well with if-statements. I think something like `passesOpcodeRequirements` would be more straightforward here. As it stands, the `enforce` name could be a bit confusing in terms of whether expected return is `true` or `false` in the success case.

Reply:
I've changed it to `satisfiesOpcodeRequirements`